modules/task/cherami/cherami.go

```diff
@@ -261,20 +269,43 @@ func (b *Backend) consumeAndExecute() {
 	}()
 	for delivery := range b.deliveryCh {
 		messageData := delivery.GetMessage().GetPayload().GetData()
-		// TODO (madhu): Only specific errors should be retried
-		// TODO (madhu): Once context is added to the message, use that here
-		ctx := context.Background()
-		if err := task.Run(ctx, messageData); err != nil {
+		b.withContext(delivery, func(ctx context.Context) {
+			// TODO (madhu): Only specific errors should be retried
+			if err := task.Run(ctx, messageData); err != nil {
+				ulog.Logger(ctx).Error("Task run failed", zap.Error(err))
+				b.taskFailure.Inc(1)
+				if err := delivery.Nack(); err != nil {
+					ulog.Logger(ctx).Error("Delivery Nack failed", zap.Error(err))
+				}
+			} else {
+				b.taskSuccess.Inc(1)
+				if err = delivery.Ack(); err != nil {
+					ulog.Logger(ctx).Error("Task ack to cherami failed", zap.Error(err))
+				}
+			}
+		})
+	}
+}
+
+func (b *Backend) withContext(delivery cherami.Delivery, f func(context.Context)) {
+	ctxData := delivery.GetMessage().GetPayload().GetUserContext()
+	ctx := context.Background()
+	var span opentracing.Span
+	if ctxVal, ok := ctxData[_ctxKey]; ok {
+		spanCtx, err := b.ctxEncoder.Unmarshal([]byte(ctxVal))
+		if err != nil {
+			ulog.Logger(ctx).Error("Unable to decode context", zap.Error(err))
 			b.taskFailure.Inc(1)
-			ulog.Logger(ctx).Error("Task run failed", zap.Error(err))
-			_ = delivery.Nack()
-		} else {
-			b.taskSuccess.Inc(1)
-			if err = delivery.Ack(); err != nil {
-				ulog.Logger(ctx).Error("Task ack to cherami failed", zap.Error(err))
+			if err := delivery.Nack(); err != nil {
```

Comment:
If we fail to acknowledge delivery/decode context, we still continue to execute task?